code/modules/projectiles/guns/energy.dm

```diff
@@ -293,6 +293,250 @@
+	overcharge_datums = list(
+						/datum/lasgun/base/standard,
+						/datum/lasgun/base/overcharge,
+						/datum/lasgun/base/heat,
+						/datum/lasgun/base/wide,
+						/datum/lasgun/base/focus,
+						/datum/lasgun/base/pulse)
+
+	flags_gun_features = GUN_AUTO_EJECTOR|GUN_CAN_POINTBLANK|GUN_AMMO_COUNTER|GUN_ENERGY|GUN_AMMO_COUNTER
+	starting_attachment_types = list(/obj/item/attachable/attached_gun/grenade, /obj/item/attachable/stock/tx73)
+	attachable_offset = list("muzzle_x" = 32, "muzzle_y" = 18,"rail_x" = 12, "rail_y" = 23, "under_x" = 23, "under_y" = 15, "stock_x" = 22, "stock_y" = 12)
+
+	accuracy_mult_unwielded = 0.5 //Heavy and unwieldy; you don't one hand this.
+	scatter_unwielded = 100 //Heavy and unwieldy; you don't one hand this.
+	damage_falloff_mult = 0.25
+	fire_delay = 3
+
+//This is where you balance the Lasgun.
+/datum/lasgun/base
+	var/charge_cost = 0
+	var/ammo = null
+	var/fire_delay = 0
+	var/fire_sound = null
+	var/message_to_user = ""
+
+/datum/lasgun/base/standard
+	charge_cost = 10
+	ammo = /datum/ammo/energy/lasgun/M43
+	fire_delay = 3
+	fire_sound = 'sound/weapons/guns/fire/laser3.ogg'
+	message_to_user = "You set the Lasgun's charge mode to standard fire."
+
+/datum/lasgun/base/overcharge
+	charge_cost = 20
+	ammo = /datum/ammo/energy/lasgun/M43/overcharge
+	fire_delay = 10
+	fire_sound = 'sound/weapons/guns/fire/laser3.ogg'
+	message_to_user = "You set the Lasgun's charge mode to overcharge."
+
+/datum/lasgun/base/heat
+	charge_cost = 20
+	ammo = /datum/ammo/energy/lasgun/M43/heat
+	fire_delay = 8
+	fire_sound = 'sound/weapons/guns/fire/laser3.ogg'
+	message_to_user = "You set the Lasgun's charge mode to wave heat."
+
+/datum/lasgun/base/wide
+	charge_cost = 20
+	ammo = /datum/ammo/energy/lasgun/M43/blast
+	fire_delay = 8
+	fire_sound = 'sound/weapons/guns/fire/laser3.ogg'
+	message_to_user = "You set the Lasgun's charge mode to wide."
+
+/datum/lasgun/base/pulse
+	charge_cost = 10
+	damage_mod = -0.15
+	gun_firemode_list_mod = list(GUN_FIREMODE_AUTOMATIC)
+	fire_delay = 2.5
+	fire_sound = 'sound/weapons/guns/fire/laser3.ogg'
+	message_to_user = "You set the Lasgun's charge mode to pulse."
+
+
+//Toggles Overcharge mode. Overcharge mode significantly increases damage and AP in exchange for doubled ammo usage and increased fire delay.
+/obj/item/weapon/gun/energy/lasgun/proc/toggle_chargemode(mob/user)
+	var/max_overcharge_mode = length(overcharge_datums)
+	if(overcharge_position >= max_overcharge_mode)
+		overcharge_position = 1
+	else 
+		overcharge_position += 1
+
+	playsound(user, 'sound/weapons/emitter.ogg', 5, 0, 2)
+	charge_cost = initial(overcharge_datums[overcharge_position].charge_cost)
+	ammo = GLOB.ammo_list[initial(overcharge_datums[overcharge_position].ammo)]
+	fire_delay = initial(overcharge_datums[overcharge_position].fire_delay)
+	fire_sound = initial(overcharge_datums[overcharge_position].fire_sound)
+	to_chat(user, initial(overcharge_datums[overcharge_position].message_to_user))
+
```

Comment:
Finally here, you would add in `icon_state = initial(overcharge_datums[overcharge_position].icon_state)`

and also `update_icon()`